src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+// Copyright 2018-2023 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/*
+ * shmem.c --
+ *
+ * This file contains the implementation for managing shared memory created
+ * for use by SplinterDB and all its innards.
+ */
+#include <unistd.h>
+
+#include "platform.h"
+#include "shmem.h"
+#include "util.h"
+
+// SplinterDB's shared segment magic identifier
+#define SPLINTERDB_SHMEM_MAGIC (uint64)0xDEFACADE
```

Comment:
Discussed and we agreed to take it out.

Reply:
@rtjohnso - I have reworked the interfaces in this file to consistently take-in `shmem_heap *` parameter.

I have also purged the `platform_heap_handle *` from this file.
All interfaces in this module now work with `platform_heap_id *hid` and `shmem_heap *`.

This brings the code closer to your wish of using the type-system to defend against bad shmem handle pointers.

But, even after this cleanup, and due to my own refactoring, the following check and error in `platform_shmdestroy()` was still tripping-up while running some unit-tests:

```
 334    if (shm->shm_magic != SPLINTERDB_SHMEM_MAGIC) {
 335       platform_error_log("%s(): Input heap ID, %p, does not seem to be a "
 336                          "valid SplinterDB shared segment's start address."
 337                          " Found magic 0x%lX does not match expected"
 338                          " magic 0x%lX.\n",
 339                          __func__,
```

I have, since, discovered and fixed this self-introduced bug in this line of code in the destroy function:

```
 320    const void *shmaddr = (const void *)platform_heap_id_to_shmaddr(*hid_out);
```

`platform_heap_id_to_shmaddr()` needed to be called by `*hid_out` and not `hid_out`.

The point being: Given that this check on SHMEM_MAGIC did come handy even during my own refactoring and stabilization, I see some value for retaining this (even with the improvements in various interfaces that now take typed arguments).

So, even thought I think you objected to this poor-man's consistency checking, I wish to retain it, so it will provided one more last-line-of-defense against code refactoring in the future.